### DISCUSSION TITLE
Action check for Caption property

### DISCUSSION THREAD

**Discussion by passion4dev** [StefanMaron/BusinessCentral.LinterCop#121] [New Rule]
I think it would be usefull if to check wether a Caption is set on an action i.e. :

         action(wCreateCustSalesPrice)
         {
                ApplicationArea = all;
                Image = PriceAdjustment;
                Caption = 'Create Sales Price';  <====

                trigger OnAction()
                var
                    wlRecSalesLine: Record "Sales Line";
                begin
                    CurrPage.SetSelectionFilter(wlRecSalesLine);
                    Report.RunModal(Report::wSalesLineCreateCustPriceCSX);
                end;
          }

**Comment by rvanbekkum** [comment]
And also for enum value captions, table field captions, group captions and other page controls where applicable.

**Comment by rvanbekkum** [comment]
And skip the check for controls where `ShowCaption = false;` is set.

**Comment by fvet** [comment]
Same should be applied on **Table** captions as well.

**Comment by StefanMaron** [comment]
This will be one of the next rule I will work on. Is this list complete?

- Action
- enum value
- table fields
- page groups
- table

**Reply by jwikman** [reply]
I just ran a check on the BaseApp to see what TransUnits in the g.xlf file that didn't have any source in the AL code, just to test the parsing capabilities of my NAB AL Tools extension.
And it seems as if the Microsoft developers need this rule! 😂
I found 600+ places in the BaseApp that was missing captions.

They where mainly:
- Page
- Query
- Table
- Table Field
- Page Field (if not a Table Field with caption)
- Actions

And then there where som missing OptionCaption on Table Fields.

So I would like to add the following to your list
- Page
- Query
- Page Field (if not a Table Field with caption)
- OptionCaption on Table Fields as well, if possible? Feels like overkill to add a new rule only for this...

**Reply by StefanMaron** [reply]
Thanks for the effort! Will dig into this this week for sure!

**Reply by StefanMaron** [reply]
What about empty Captions?
![image](https://user-images.githubusercontent.com/24838311/155271521-4209d828-4dad-476c-af1e-8e7d39f68357.png)

**Reply by rvanbekkum** [reply]
I think that should be allowed (you often want `Locked = true`, but I rhink that might be better left to the developer to decide)

**Reply by jwikman** [reply]
I agree. An empty caption is also a caption and should not result in a warning. 👍

**Comment by jwikman** [comment]
Thanks @StefanMaron, this rule seems to do it's job!
I just found a couple of warnings for both LC0016 and LC0017 that obviously was missed during code reviews...
Well done! 👍

**Comment by StefanMaron** [comment]
I think we will also need to consider captionclass right?

**Reply by jwikman** [reply]
Yes, didn't think of the CaptionClass. But you've solved that already, right? I'm pretty sure I saw the warning earlier today!?🤔

This rule also have a couple of other "false positives that I got today (depending on how you look at it)
- If the SourceTable of a page is a system table (like the `Field` table), the field always has a caption, so no need to throw warnings here
- Caption on API objects (query and page). Captions should probably be added on fields, but not needed on object level. (this is how MS does it)

**Reply by StefanMaron** [reply]
Captionclass is already fixed. 
Thanks for pointing out the others as well. I will look into them.
